src/Application/Actions/Donations/Confirm.php

```diff
@@ -244,4 +259,31 @@ public function randomString(): string
     {
         return substr(Uuid::uuid4()->toString(), 0, 15);
     }
+
+    private function handleCardException(
+        string $context,
+        CardException $exception,
+        Donation $donation,
+        string $paymentIntentId,
+    ): JsonResponse {
+        $exceptionClass = get_class($exception);
+        $this->logger->info(sprintf(
+            'Stripe %s on Confirm %s for donation %s (%s): %s',
+            $exceptionClass,
+            $context,
+            $donation->getUuid(),
+            $paymentIntentId,
+            $exception->getMessage(),
+        ));
+
+        $this->entityManager->rollback();
```

Comment:
I think I'd move the rollback call outside, e.g.:

```
$this->entityManager->rollback();
 return $this->handleCardException(
            $this->logger->info(sprintf(
                context: 'confirmPaymentIntent',
                'Stripe %s on Confirm for donation %s (%s): %s',
                exception: $exception,
                $exceptionClass,
                donation: $donation,
                $donation->getUuid(),
                paymentIntentId: $paymentIntentId,
                $paymentIntentId,
            );
```

It's probably a good rule of thumb that `beginTransaction`  and `rollback` should both be called from the same function.

Reply:
https://github.com/thebiggive/matchbot/pull/726/commits/3ab0d9aa83521c10db85209839a48587ba9b1d95